File: src/ts/plugins/pluginOwner.ts
/**
 * Every stored plugin value belongs to the plugin whose `//@name` banner matches
 * its owner. Values that arrived from an upstream RisuAI save without an
 * ownership sidecar carry the sentinel below. The banner is parsed line by line,
 * so a plugin name can never contain NUL and plugin code cannot forge it.
 */
export const UNOWNED_PLUGIN_OWNER = '\u0000unowned'

export function isUnownedPluginOwner(owner: string): boolean {
    return owner === UNOWNED_PLUGIN_OWNER
}

export function isValidPluginOwner(owner: string): boolean {
    if (owner === UNOWNED_PLUGIN_OWNER) return true
    return owner.length > 0 && owner.length <= 512 && !owner.includes('\u0000')
}

/** Ownership as an upstream save carries it beside the flattened values. */
export interface PluginStorageMetaEntry {
    plugin: string
    updatedAt: number
}

export type PluginStorageMeta = Record<string, PluginStorageMetaEntry>

export function readPluginStorageMetaOwner(
    meta: PluginStorageMeta | undefined,
    key: string,
): string {
    const entry = meta?.[key]
    if (!entry || typeof entry.plugin !== 'string') return UNOWNED_PLUGIN_OWNER
    if (isUnownedPluginOwner(entry.plugin) || !isValidPluginOwner(entry.plugin)) {
        return UNOWNED_PLUGIN_OWNER
    }
    return entry.plugin
}

File: src-tauri/src/persistent_store/plugin_owner.rs
//! Plugin storage ownership.
//!
//! Every `plugin_storage` row belongs to the plugin whose `//@name` banner
//! matches `owner`. Rows that arrived through an upstream RisuSave without an
//! ownership sidecar carry the sentinel below. The banner is parsed line by
//! line, so a real plugin name can never contain NUL and the sentinel cannot be
//! forged from plugin code.

/// Owner of a row whose plugin is not known.
pub(crate) const UNOWNED_OWNER: &str = "\u{0000}unowned";

/// `claimed_from` value written when an initialization session moves a row away
/// from the sentinel.
pub(crate) const CLAIMED_FROM_UNOWNED: &str = "unowned";

pub(crate) fn is_unowned(owner: &str) -> bool {
    owner == UNOWNED_OWNER
}

/// Rejects names that could collide with the sentinel or overflow a key
/// component. Callers validate before writing so a forged owner cannot enter
/// the table through an import path.
pub(crate) fn validate_owner(owner: &str) -> bool {
    if owner == UNOWNED_OWNER {
        return true;
    }
    !owner.is_empty() && owner.len() <= 512 && !owner.contains('\0')
}
